### PR TITLE
add auto voicechannel feature

### PR DESCRIPTION
**Pull request by TomRomeo**
Hello! I really like this bot and wanted to contribute a feature. 😊
When I was checking out the discussions for possible features to implement, I came across the auto voicechannel feature.

---
## Short description of the feature:
Some voicechannels can be set as 'auto voicechannels'.
When a member joins this channel, a custom voicechannel will be created for them.
When they leave this channel, it gets removed again.

---

## For this feature I have implemented the following commands:

- `autovc  add #voicechannel`
- `autovc  remove #voicechannel`
- `autovc purge`
- `autovc show`

and the following permissions:

- `sp.guild.config.autochannel`
- `sp.chat.autochannel`

---

## The current behavior is the following:

- When someone joins an auto voicechannel, they get moved to a channel named after themselves.
- When they move or leave out of that channel, it gets deleted again.
- When they join the auto voicechannel from their custom vc, the custom vc gets removed but no new one is created.

# Some things that might be nice to add:

- A possibility to rename the custom vcs might be nice? Or a prefix for custom vcs
- Some commands to lock the channel to a number of members or to add a whitelist might be nice
- Right now, the channels would not get removed when the bot goes offline, that is probably no great

Let me know what you think! 👋

### CODE REVIEW

**Comment by zekroTJA** [comment]
Man, that looks awesome. Thank you very much for the contribution! 😍

I will test around a bit with your implementation later today, but looks very good as what I've seen already.

File: internal/commands/cmdautovc.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/zekroTJA/shinpuru/internal/util"
+	"github.com/zekroTJA/shinpuru/internal/util/static"
+	"github.com/zekroTJA/shinpuru/pkg/acceptmsg"
+	"github.com/zekroTJA/shinpuru/pkg/fetch"
+	"github.com/zekroTJA/shinpuru/pkg/stringutil"
+	"github.com/zekroTJA/shireikan"
+	"github.com/zekrotja/dgrs"
+)
+
+type CmdAutovc struct {
```
Comment (zekroTJA) [inline internal/commands/cmdautovc.go:19]: Awesome that you also provided an implementation for the legacy command system, but you forgot to register the command in the handler. ^^

Just add the following line to `internal/inits/legacycommandhandler.go`
```
cmdHandler.RegisterCommand(&commands.CmdAutovc{})
```

File: internal/listeners/autovc.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	"github.com/zekroTJA/shinpuru/internal/services/database"
+	"github.com/zekroTJA/shinpuru/internal/services/permissions"
+	"github.com/zekroTJA/shinpuru/internal/util/static"
+	"github.com/zekrotja/dgrs"
+	"log"
+	"strings"
+)
+
+var autovcCache = map[string]string{}
+var voiceStateCache = map[string]*discordgo.VoiceState{}
```
Comment (zekroTJA) [inline internal/listeners/autovc.go:15]: Would make sense to add these two map instances to the `ListenerAutoVoice` struct. Doesn't make a huge difference in the logic and behavior, but 1) I am not really a fan of singleton dependencies and 2) I did it like that in every other listener and would also change it here to keep consistency. :)

File: internal/listeners/autovc.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+func (l *ListenerAutoVoice) createAutoVC(s *discordgo.Session, userID, guildID string) error {
+	user, err := l.st.User(userID)
+	if err != nil {
+		return err
+	}
+	ch, err := s.GuildChannelCreate(guildID, user.Username, discordgo.ChannelTypeGuildVoice)
+	if err != nil {
+		return err
+	}
```
Comment (zekroTJA) [inline internal/listeners/autovc.go:97]: This implementation works for voice channels which are in the root of the guilds channel list. When the voice channel is in a channel category, the new voice channel should also be added as child for the category and re-positioned after to be below the auto voice channel.

Discord_E70EMvoaHK

File: internal/services/database/mysql/migrations.go [hunk 4]
```diff
@@ -119,3 +120,11 @@ func migration_10(m *sql.Tx) (err error) {
...
 	return
 }
+
+// VERSION 11:
+// - add property `autovc` to `guilds`
+func migration_11(m *sql.Tx) (err error) {
+	err = createTableColumnIfNotExists(m,
+		"guilds", "`autovc` text NOT NULL DEFAULT ''")
+	return
+}
```
Comment (zekroTJA) [inline internal/services/database/mysql/migrations.go:130]: This implementation only allows one auto voice channel for each guild. Would be better if we create a new table where all created auto VCs are listed in, so you can create multiple per guild.

The table would be like
| autovcs |
|---|
| iid `int` *(PK - auto increment)* |
| guildID `string` |
| chanID `string` |

Wouldn't include this as requirement for this PR and would later modify it, if you don't want to do it. ^^

**Review by zekroTJA — CHANGES_REQUESTED** [review]
Found some things that might need a change, if you don't mind. ^^

If you have any questions or need some help for something, feel absolutely free to contact me on my [Discord](https://dc.zekro.de) (`zekro#0001`) or [Twitter](https://twitter.com/zekrotja) or simply via [mail](mailto:contact@zekro.de). :)

**Comment by TomRomeo** [comment]
Hey! Thanks for the thorough review! 😄I will be working on your suggestions!

File: internal/services/database/mysql/migrations.go (hunk 4 above)
Comment (TomRomeo) [reply internal/services/database/mysql/migrations.go:130]: Setting multiple auto VCs should actually work.
When implementing the autovc feature, I closely followed the implementation of the `autorole` command.
There the roles get concatenated to a string separated by semicolons.
That said, a separate table for auto VCs might be the cleaner way to implement this.
Let me know what you think and I can implement it this way 😄

File: internal/listeners/autovc.go (hunk 3 above)
Comment (TomRomeo) [reply internal/listeners/autovc.go:100]: Oh that's true, I totally missed that. 😅
I've added this behavior with 366d46a.
Is it correct this way?

File: internal/listeners/autovc.go (hunk 2 above)
Comment (TomRomeo) [reply internal/listeners/autovc.go]: Yeah, that makes sense,
I've changed it in fefc822, is it okay that way?

Comment (zekroTJA) [reply internal/listeners/autovc.go]: Yep, that looks fine.

File: internal/services/database/mysql/migrations.go (hunk 4 above)
Comment (zekroTJA) [reply internal/services/database/mysql/migrations.go:130]: Ah okay, sorry, I missed that. Then I would let the implementation this way for now. 

Sadly, my database implementation is very old and also extremely unoptimized and bad implemented. Maybe later I find the time to completely rework the database bindings, but that would also need a lot migration work. ^^

File: internal/listeners/autovc.go (hunk 3 above)
Comment (zekroTJA) [reply internal/listeners/autovc.go:101]: Just tested around a bit and works great. Thank you. ^^

**Review by zekroTJA — APPROVED** [review]

**Comment by zekroTJA** [comment]
Looks very goo.d Thank you for your awesome contribution. ^^

**Pull request merged**